Boost default tick rate to double digit kHz on tickless hardware

andyross opened the pull request:
Since the introduction of pervasively-supported tickless mode, it's been possible for apps to achieve effectively arbitrary timing precision (up to the limits of the hardware counter, of course) by adjusting the CONFIG_SYS_CLOCK_TICKS_PER_SEC value.

But experience has shown that this mechanism is too confusing or too controversial for app authors, its uptake has been very low, and there are now currently two separate reimplementations of system timing architectures in review.  All to achieve a feature that... was supposed to already be present.

So let's just make it default.

On all tickless capable hardware platforms, the default tick rate is now 10 kHz.  On nRF5x devices, which have a counter that doesn't divide well into decimal frequencies, we're using a raw tick rate of 32768 Hz to exactly match the hardware counter (pleasingly, this GREATLY simplifies the generated code for the ISR and should show significantly improved latency numbers).  Note that software emulation platforms (qemu and nsim) which expose the host clock must continue to use a 100 Hz default, as the emulator timing precision is limited by host OS scheduling quanta.

Note that this is a small patch series (that removes more lines than it adds!), and that almost all the content is just kconfig muckery. There were only a handful of bugs fixed, two were bad test assumptions about tick rate and one was an issue found by inspection in the "next tick round up" code in the nrf driver.

This completes sanitycheck correctly on the emulation platofrms, and runs tests/timer and tests/sched tests successfully on nrf52840_pca10056 and frdm_k64f.  I'm having trouble with my SAM E70 board at the moment but hope to have results from that in a bit.

Anyway, hopefully this can help people looking for increased precision in the near term while we continue to discuss the various rewrites.

Commit:
drivers/timer/nrf_rtc_timer: Fix round-up for rapid tick rates

When the tick rate was less than MIN_DELAY, bumping a "too soon"
expiration by just one tick may not be enough and we could
theoretically miss the counter.

Instead, eliminate the MIN_DELAY computation and write to the spec:
NRF guarantees that the RTC will generate an interrupt for a
comparator value two cycles in the future.  And further, we can test
at the set point to see if we "just missed" the interrupt (i.e. zero
cycles delay) and flag a synchronous interrupt.  So we only need to
miss a requested interrupt now for the special case of exactly one
cycle in the future, and then we're only late by one cycle.  That's
optimal.

Also fixes an off-by-one in the next cycle computation.  By API
convention, an ticks argument of one or less means "at the next tick"
and not "right now".  So we need to add one to the target cycle to
avoid incorrectly triggering a synchronous interrupt.  This was a
non-issue when a tick is longer than a hardware cycle but is needed
now.

Also handles the edge case with zero latency interrupts (which are
unmaskable) which might mess up timing.  This was always a problem,
but we're more sensitive now and it's comparatively more likely to
occur.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/mem_pool: Fix timeout units

This code was clearly written to assume that the timeout argument to
k_mem_pool_alloc() was in ticks and not ms.  Adjust to what appears to
have been the intent.  It was working as intended (i.e waiting one or
1/10th of a second) only on systems where the default tick rate was
100 Hz.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests: samples: Apps shouldn't set tick rate

Tick rate is becoming a platform tunable in the tickless world.  Some
apps were setting it due to requirements of drivers or subsystems (or
sometimes for reasons that don't make much sense), but the dependency
goes the other way around now: board/soc/arch level code is
responsible for setting tick rates that work with their devices.

A few tests still use hard-configured tick rates, as they have
baked-in assumptions (like e.g. "a tick will be longer than a
millisecond") that need to be addressed first.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Crank up default tick rate

When tickless is available, all existing devices can handle much
higher timing precision than 10ms.  A 10kHz default seems acceptable
without introducing too much range limitation (rollover for a signed
time delta will happen at 2.5 days).  Leave the 100 Hz default in
place for ticked configurations, as those are going to be special
purpose usages where the user probably actually cares about interrupt
rate.

Note that the defaulting logic interacts with an obscure trick:
setting the tick rate to zero would indicate "no clock exists" to the
configuration (some platforms use this to drop code from the build).
But now that becomes a kconfig cycle, so to break it we expose
CONFIG_SYS_CLOCK_EXISTS as an app-defined tunable and not a derived
value from the tick rate.  Only one test actually did this.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
boards/arm: Remove 1000 Hz tick rate on STM32 boards

These all have what appears to be a promiscuously cut-and-pasted
declaration for a 1000 Hz tick rate.  They are all SysTick boards and
will work very well with the new 10 kHz default, so use that instead.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
boards/arc: Remove tick rate settings for ARC hardware

The ARC timer is a MHz-scale cycle counter and works very well with
the new 10 kHz default tick rate.  Remove the settings for ARC
hardware.

Note that the nsim board definitions are left at 100 Hz.  That is a
software emulation environment that (like qemu) exposes the host clock
as "real" time and thus is subject to clock jitter due to host
scheduling.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/arm: Increase nRF timer default to the cycle rate.

The nRF timer runs at only 32 kHz, so there's little reason to try to
divide it to get a synthesized tick rate.  Just use the raw clock as
the tick rate, which provides maximal precision and very
singnificantly simplifies the generated code for the ISR.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc/arm: Remove 1 kHz tick rate default for mcimx7_m4 and msp432p401r

As with the STM32 boards, these had an existing default for tick rate
that is now lower than the 10 kHz default.  They're SysTick devices
that can handle the higher rate just fine.  Use that.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/sched/schedule_api: Fix slice time test for fast ticks

When ticks are sub-millisecond, the math produces minimum and maximum
values for the slice duration test that are equal.  But because of
aliasing across tick boundaries, it's always possible (for any time
period, nothing specific to time slicing here) to measure one tick
more than an intended duration.  So make sure there's always at least
a range of 1ms.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/common: Fix uptime delta test for fast ticks

The test for the k_uptime_delta utilities was calling it in a loop and
waiting for the uptime to advance.  But the code was specifically
wanting it to advance 5ms or more at one go, which clearly isn't going
to work for a tick rate above 200 Hz.

The weird thing is that the test knew this and even commented about
the limitation.  Which seems silly: it's perfectly fine for the clock
to advance just a single millisecond.  That's correct behavior too.
Let's test for that, and it will work everywhere.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/context: Fix sleep test for fast ticks

The sleep test was checking that the sleep took no longer than "2
ticks" longer than requested.  But "2 ticks" for fast tick rate
configurations can be "zero ms", and for aliasing reasons it's always
possible to delay for 1 unit more than requested (becuase you can
cross a millisecond/tick/whatever boundary in your own code on either
side of the sleep).  So that "slop" value needs to be no less than
1ms.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/sensor/hp206c: Clarify tick rate warning

The default tick rate is now 10 kHz, but that driver was demanding
that it be exactly 1 kHz instead of at least that rate.  I checked the
source, and the driver isn't actually extracting "ticks" from the
kernel illegally, it just needs fine-grained timers that work with the
existing millisecond API.  Let it build, this is fine.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
@nashif discovered two more tests with assumptions that break with fast ticks, and split out the header change that @erwango pointed out had been incorrectly squashed.

pizi-nordic commented:
Here is results on the https://github.com/zephyrproject-rtos/zephyr/pull/16653/commits/eb5dd1970ade5125be34dd21907ca364d4f9052e benchmark run on top of this PR: 
[log-proposal-32k.txt](https://github.com/zephyrproject-rtos/zephyr/files/3285409/log-proposal-32k.txt)

pizi-nordic commented:
Here is comparison of the https://github.com/zephyrproject-rtos/zephyr/commit/eb5dd1970ade5125be34dd21907ca364d4f9052e benchmark results ran on the top of #16653 and #16782 (to be more specific: I am comparing https://github.com/zephyrproject-rtos/zephyr/files/3261227/log-proposal-32k.txt with https://github.com/zephyrproject-rtos/zephyr/files/3285409/log-proposal-32k.txt)

k_timer() benchmark:

Measurement|#16653 | #16782 
-|-|-
Minimum difference between requested and measured timeout | 0 us | 7 us
Average difference between requested and measured timeout | 21.8 us | 553.9 us
Maximum difference between requested and measured timeout | 59 us | 968 us

k_usleep() benchmark:

Measurement|#16653 | #16782 
-|-|-
Minimum difference between requested and measured timeout | 30 us | 56 us
Average difference between requested and measured timeout | 50.4 us | 541.6 us
Maximum difference between requested and measured timeout | 88 us | 1023 us

joerchan commented:
@pizi-nordic Are you suggesting that we do #16653 instead of this PR?

pizi-nordic commented:
@joerchan: We need precise timing in Zephyr. We have 3 different approaches on the table. Now we should select the one most suitable for us.

The #16653 is the best in my opinion, but since I am author of this PR, this statement might be not objective :)

andyross commented:
Can we **not** do the dueling rewrite thing in this PR?  This is just a kconfig thing that changes defaults, with some fixes to tests that you're going to need for any high resolution timer implementation.

pabigot commented:
I don't have spare Zephyr time to test this; if all tests pass I won't object.  #12542 presents some concerns that may be relevant and describes a test that could reproduce them.

I would like to see the commit description of 86ad9bda5416380a184414cd9473d36530641801 changed:

    (Though in practice, the MIN_DELAY value of 32 seems very
    conservative.  Simple testing seems to show that the programming
    always happens within a single ~30us hardware cycle, so maybe
    MIN_DELAY of 2 -- the theoretical minimum to prevent slippage -- would
    be a better choice.)

32 is definitely far too conservative, but 2 won't work.  The delay between reading the counter and setting the compare register may usually be less than the 30.6 us interval of the RTC but the counter may still increment during that interval, and you'd set a compare to N+2 when the counter is at N+1 and miss.

If you're feeling lucky you could do 3.  I have doubts that the nRF51 (at 16 MHz with non-constant-time software divide) will always be take than 30.6 us (I remember being burned on #12605), so I use 4 in my RTC-based timer implementations.

Yes, this is only a change to a commit message and has no effect on the implementation, but a future maintainer may see the statement above, believe it, and replace the 32 with 2 resulting in a system that will nondeterministically experience 512 s delays.

andyross commented:
(Actually to not take my own advice: I'm convinced both @pizi-nordic's high maxima numbers and @pabigot's point about the MAX_DELAY are in fact the same root cause.  Those maximum deltas seen on current code are, in fact, just about exactly the 977us we would expect to see for a MAX_DELAY of 32.  I'm going to have to add some better optimized overrun detection to keep up with @pizi-nordic's driver, it seems.)

pizi-nordic commented:
@andyross: I used slightly different approach: If the next announcement is closer than XX us, I am not reprogramming the counter, but wait for announcement instead. I know, that this is not the best solution, but this way the counter interrupts are not being delayed even if the timeout list is constantly updated (in current implementation you can delay the clock announcement indefinitely due to MAX_DELAY).

Please also note, that @nordic-krch currently is adding overrun detection to counter API, and this will solve both your and my problem.

pabigot commented:
I don't think there's any problem with the existing overrun algorithm in `nrf_rtc_timer`; just drop `MAX_DELAY` to 4 and it should be fine.

There are tricks you can play to preserve the compare event when the next alarm's deadline has already been reached at the point it's scheduled, which eliminates `MAX_DELAY` and causes the interrupt to re-execute immediately, but they're relevant only if you have late-to-set alarms (not expressible in upstream Zephyr) or the ISR execution takes more than one tick.

andyross commented:
Getting the rollover detection right was more subtle than expected, and once working it exposed what seems to be a nRF HAL bug (filed in in #16823, but worked around in this PR).  But this should be close to optimal precision as far as "how close to now can we set a timeout".

Commit:
kernel/sched: Interpret zero timeslice time correctly

The scheduler API has always allowed setting a zero slice size as a
way to disable timeslicing.  But the workaround introduced for
CONFIG_SWAP_NONATOMIC forgot that convention, and was calling
reset_time_slice() with that zero value (i.e. requesting an immediate
interrupt) in circumstances where z_swap() had been interrupted
nonatomically.

In practice, this never happened.  And if it did, it was a single
spurious no-op interrupt that no one cared about.  Until it did,
anyway...

Now that ticks on nRF devices are at full 32 kHz speed, we can get
into a situation where the rapidly triggering timeslice interrupts are
interrupting z_swap() calls, and the process feeds back on itself and
becomes self-sustaining.

Put that test into the time slice code itself to prevent this kind of
mistake in the future.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/sleep: Fix usleep test for fast ticks

The logic about minimal sleep sizes due to "tick" aliasing was
correct, but drivers also have similar behavior with "cycle" aliasing
too.  When cycles are 3-4 orders of magnitude faster than ticks, it's
undetectable noise.  But now on nRF they're exactly the same and we
need to correct for that, essentially doubling the number of ticks a
usleep() might wait for.

The logic here was simply too strict, basically.  Fast tick rates
can't guarantee what the test promised.

Note that this relaxes the test bounds on the other side of the
equation too: it's no longer an error to usleep() for only one tick
(i.e. an improved sleep/timeout implementation no longer gets detected
as a test failure).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/tickless/tickless_concept: Force 100 Hz ticks

This test was written to assume ~100 Hz ticks in ways that are
difficult to fix.  It wants to sleep for periods on the order of the
TICKLESS_IDLE_THRESH kconfig, which is extremely small on high tick
rate systems and (on nRF in particular) does not have a cleanly
divisible representation in milliseconds.

Fixing precision issues by cranking the idle threshold up on a
per-system basis seems like an abuse, as that is what we want to be
testing in the first place.  Just let the test run at the tick rate it
has always expected.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/early_sleep: Fix for fast ticks

"50" ticks is fine with 100 Hz timer precision but way too short to
survive the conversion to milliseconds on fast, non-decimal tick
rates.  Make it half a second, which was the original intent.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/workq/work_queue: Fix for fast/non-standard tick rates

This test was written to properly align its millisecond-measured wait
time and assumed that there would be no other overhead.  In fact on
fast tick rate systems (or even ones where the alignment computation
doesn't provide the needed padding as "slop") that's not quite enough
time to complete the full test.  There are cycles between the sleep
calls that need to be accounted for, and aren't.

Just give it one extra work item of time before failing.  We aren't
testing work queue timing precision here, just evaluation semantics.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/posix/common: Adjust miscalibrated timing test

This test seems a little confused.  It does a POSIX usleep() for 90ms,
then checks the time taken, and verifies that it was no less
than... 91ms!

On existing platforms, tick alignment makes sure that we always take a
little longer, so this passes. But on high tick rate configurations we
get it exactly right.  And fail.

Adjust the calibration to allow (exactly) 90ms sleeps.  Also fixed a
comment that described the wrong units.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/cmsis_rtos_v1: Correct timing assumptions

This test was written to assume that k_busy_wait() and CMSIS
osKernelSysTick() (which is just k_cycle_get_32()) were perfectly
synchronized.  On nRF, they aren't (one is the 32 kHz RTC timer, the
other is a calibrated spin loop using the CPU frequency).

When ticks were being reported at 100 Hz granularity, there wasn't
enough precision to detect the mismatch.  Now there is.  Rework the
test to require that the clocks match to within 1%.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
CMSIS v2: Work around time unit confusion

The current CMSIS v2 implementation is clearly assuming that timeout
arguments being passed to e.g. osDelay() are in units of Zephyr ticks,
not milliseconds as specified by ARM or (inconsistently) assumed by
our test code.

Most tests work with the ~100 Hz default tick rate, but they tend to
fail on precision issues at higher tick rates.  Force the CMSIS v2
applications to be 1000 Hz for now as a workaround, and detect the
mismatch as a build failure.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/mempool: Fix ticks/ms confusion

The mempool blocking implementation was mixing tick and millisecond
APIs.  Get it right.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
So. many. tests. to fix.

I think this is set now.  It passes everything in emulation (where in theory nothing has changed) and now runs successfully on nrf52840_pca10056 across the entirely test suite (modulo the 8 failures that are already present in master).  The nrf driver is about as optimized as it's going to get (though there are alignment cycles to be saved in k_sleep and the timeout code still, I think).

And amusingly there were only two bugs fixed in the core kernel: one simple units glitch in mem_pool, and one somewhat more mindbending interaction in the scheduler between the nonatomic swap workaround, the timeslicing configuration API, and fast ticks.

pizi-nordic commented:
Here are benchmark results on the updated PR.

k_timer() benchmark:

Difference between requested and measured timeout | master | #16398 | #16653 | #16782 
-|-|-|-|-
Minimum | -5189 us | 0 us | 0 us | 0 us
Average | 435.9 us | 23.6 us | 21.8 us | 24.9 us
Maximum | 7043 us | 59 us | 59 us | 60 us 

k_usleep() benchmark:

Difference between requested and measured timeout | master | #16398 | #16653 | #16782 
-|-|-|-|-
Minimum | 10179 us | 61 us | 30 us | 30 us 
Average | 11818.4 us | 112.0 us | 50.4 us | 55.9 us 
Maximum | 11856 us | 205 us | 88 us | 85 us 

Notes:
- Numbers presented for #16653 include overhead caused by counter abstraction as well as counter-generic layer.
- In order to compare apples with apples, the CONFIG_SYS_CLOCK_TICKS_PER_SEC has been set to 32768 for #16398.
- #16398 is not intended to improve timing of the OS API. Instead it proposes new k_alarm API, which is used to as clock source for kernel.

pizi-nordic commented on tests/kernel/workq/work_queue/src/main.c:
```diff
@@ -164,6 +169,7 @@ static void resubmit_work_handler(struct k_work *work)
 		ti->key++;
 		TC_PRINT(" - Resubmitting work\n");
 		k_work_submit(work);
+	} else {
```
Why this is needed?

andyross commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -117,19 +120,39 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 
 	/* Round up to next tick boundary */
-	cyc = ticks * CYC_PER_TICK + counter_sub(t, last_count);
+	cyc = ticks * CYC_PER_TICK + 1 + counter_sub(t, last_count);
```
It's there for correctness.  If you pass in a ticks argument of zero, the API means "at the next tick" and not "immediately".   When ticks were longer than cycles, this didn't matter.  Now that we can trigger a synchronous interrupt, the distinction matters and this is here to avoid early timeouts.   Added a note in the commit message about it.

andyross commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -63,7 +61,10 @@ void rtc1_nrf_isr(void *arg)
 		u32_t next = last_count + CYC_PER_TICK;
 
-		if (counter_sub(next, t) < MIN_DELAY) {
+		/* As below: we're guaranteed to get an interrupt as
```
This could happen to the previous implementation too, it's nothing new.  And it's fundamental with any kind of comparator architecture like this.  If you can arbitrarily preempt the timer driver it's always possible to mess up its logic.

Stated more simply: if you need good realtime timing, you can't be preempting the driver that provides that.

But one good workaround for an app that aboslutely had to use a ZLI or NMI (or something at a higher priority than timer) would be to chain to a lower priority exception (like PendSV) that can then flag a timer interrupt synchronously, which will reset the state.  You could have a tiny high priority ("MetaIRQ") thread for exactly this purpose.  With a little work you could optimize this to recognize if a timer ISR had been interrupted and skip the extra step.

andyross commented on kernel/sched.c:
```diff
@@ -255,9 +255,10 @@ static void reset_time_slice(void)
-	_current_cpu->slice_ticks = slice_time + z_clock_elapsed();
-
-	z_set_timeout_expiry(slice_time, false);
+	if (slice_time != 0) {
```
We can't, because the time slicing is specific to a single CPU in a SMP system and not global like a timeout.

pizi-nordic replied:
What about per-cpu slicing timeout?

pizi-nordic replied:
What about per-cpu slicing timeout?

pabigot replied:
Could you expand more on your question?  I don't understand what you're asking for/about.

pizi-nordic replied:
At the moment the thread time slicing and timers are both using the same hardware providing system clock. So for example, when you are setting the timeout you have to check both slicing and timeout list to check when next timer interrupt will happen. My question is if we could implement slicing as another user of timeout list instead of bolting it on a side.

pabigot replied:
Thanks.  So it's like the inverse of what I did in k_alarm, where the system clock was just another alarm.

My view: (a) this might be worth pursuing, but (b) it doesn't belong in and should not block this PR.  I'd put it into a new issue where it could be discussed properly.

pizi-nordic replied:
Yes, this does not belong here (this is why this thread starts with "Side question"). I just wanted to hear Andy's opinion.

andyross replied:
Using a separate counter and interrupt for timeslicing would be simpler for sure, but not all hardware has one.  And obviously seizing a device for timeslicing means that it won't be available for an app or driver to use.

Right now the architecture is that the kernel wants a single timer driver and it multiplexes it internally for use by timeouts and time slicing.  I could see value in a scheme where the kernel had separate "drivers" for those two tasks and the driver layer would do the multiplexing on hardware where they were the same underlying device.

pizi-nordic replied:
I am not proposing using separate timer. I just would like to setup slicing as regular entry in timeout list. If the thread will be swapped out, such entry will be removed (and readded of the next thread should be sliced too). Otherwise the kernel will be notified through callback about the need of slicing.

On the SMP this might be a bit tricky, as we will have to use IPI to forward the slicing request to the correct CPU as slicing is per-cpu.

BTW: @andyross How slicing is currently implemented on SMP targets? Are you using separate timers with interrupts bound to given CPU?

andyross commented:
Rebased, removed the debug cruft @pizi-nordic pointed out.  Clarified comments in the nrf patch.

andyross commented:
I found an extra cycle to save, just pushed.  The case where we have an imminent interrupt has three subcases, not two: (1) we missed it and can flag an interrupt immediately, (2) we have to slip by a cycle, **but also** (3) the newly set timeout is still two cycles in the future and we can leave it in place.   This morning we were slipping (3) also, needlessly.  Worth a few microseconds anyway.

And, while @pizi-nordic has the rig ready, if you're willing to spin here (as I think the new implementation does, right?) you can do even better.  In case (2) we can spin to the next cycle instead of slipping, fixing the one overrun case.  Personally I don't think this is a good idea for the reasons detailed in the comment, but if that's the competition and people agree we want it, it's probably worth measuring:

```diff
--- a/drivers/timer/nrf_rtc_timer.c
+++ b/drivers/timer/nrf_rtc_timer.c
@@ -20,6 +20,23 @@
                      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
 #define MAX_TICKS ((COUNTER_MAX - CYC_PER_TICK) / CYC_PER_TICK)
 
+/* Experimental feature.  We have the capability of detecting the case
+ * where an interrupt is to be set "at the next cycle", but no ability
+ * in the hardware to deliver that interrupt.  The default choice is
+ * just to slip the event to the next cycle. But... for some apps,
+ * maybe it's preferable to spin instead for accuracy.
+ *
+ * Note that this has severe tradeoffs: one of course is that we're
+ * increasing latency for every other interrupt in the system,
+ * potentially by a lot (e.g. on a 64 MHz nrf52, a RTC cycle is 2000
+ * CPU cycles).  A more subtle one is that it introduces priority
+ * inversion conditions, where it's now possible for a high priority
+ * thread to "sleep" expecting to give the CPU to a lower priority
+ * thread, yet because of the spin here the thread won't actually pend
+ * and the lower priority thread will never run.
+ */
+#define SPIN_FOR_IMMINENT_INTERRUPT 1
+
 static struct k_spinlock lock;
 
 static u32_t last_count;
@@ -162,7 +179,13 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
                        NVIC_SetPendingIRQ(RTC1_IRQn);
                } else if (counter_sub(cyc, t2) == 1) {
                        /* Too soon, interrupt won't arrive. */
+#ifdef SPIN_FOR_IMMINENT_INTERRUPT
+                       while (counter_sub(cyc, counter()) == 1) {
+                       }
+                       NVIC_SetPendingIRQ(RTC1_IRQn);
+#else
                        set_comparator(cyc + 1);
+#endif
                }
                /* Otherwise it was two cycles out, we're fine */
        }
```

pabigot commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -117,19 +120,39 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+		if (cyc == t2) {
+			NVIC_SetPendingIRQ(RTC1_IRQn);
+		} else {
+			set_comparator(cyc + 1);
```
But I have to agree with @pizi-nordic about this. `t` is no longer relevant; it's fine.  But what's preventing `COUNTER` from incrementing between the read of `t2` and the set of the comparator on the next line?  The fact that `cyc-t2` is at least 2 does not appear to guarantee that at the instant `COMPARE` was actually set `COUNTER` was *still* 2 below `cyc`.

Does something like this make sense?

```
t2 = counter();
set_comparator(cyc);
unsigned delta = counter_sub(cyc, t2);
if (delta == 0) {
   // Missed it but deadline has been reached, can force IRQ
} else if (delta <= 2) {
   // Maybe missed it; deadline hasn't been reached so can't force IRQ.
   // Set to next counter value guaranteed to be in time.
   set_comparator(t2 + 3);
} else {
   // At least 2 cycles out, we're fine
}
```

pizi-nordic replied:
IMHO we should check the current state of the counter AFTER we set compare value. If the current value is ahead of the alarm we are safe (ahead = 2 ticks or more), otherwise we either set the alarm and was pre-empted (we can check this checking IRQ pending bit) or we missed the alarm.

pabigot reviewed (changes requested):
I like this.  A lot.  Initial tests (Nordic only) seem to have great performance.  If this continues to show well and gets merged I can discard #16398 and address my alarm behavior needs in a different way.  (Caveat: only done basic testing, but I intend to merge this into my development branch so may have more insights in a week or so.)

I've added one specific comment describing a case on Nordic where I too believe the two-advance minimum requirement for setting compare could be missed, which I'd like resolved before approving (modify to eliminate risk, or explain why it can't happen).

We're gonna need `k_cycle_get_64()`.  We don't need it in this PR.

A high level concern: if somebody changes `CONFIG_SYS_CLOCK_TICKS_PER_SEC` to a non power of two, re-introducing call to execute an integer divide, I'm not convinced the nRF51 will meet the tight deadlines this implementation imposes. Personally I'd be ok with a compile-time assert that prevents that sort of (mis)configuration, but the Nordic folks should approve that.

I am strongly against `SPIN_FOR_IMMINENT_INTERRUPT`.

I am mildly against `NVIC_SetPendingIRQ(RTC1_IRQn)` to force for a deadline that's been reached.  Yes, it works.  However, unlike attempts to [preserve the `CC[0]` event as in #16398](https://github.com/zephyrproject-rtos/zephyr/pull/16398/files#diff-2bfa0ece3c3dfc742c1fe2bc0ebe1822R394) this approach requires that the IRQ handler be restricted to a single compare channel.

There are two other channels on the same timer which could be profitably used in certain cases.  In one the Bluetooth stack could use a second compare register on the system clock to do its coarse timing, freeing up a whole RTC peripheral instance for application use.  In a related situation a second or third channel could be used to capture the system clock at specific events (e.g., radio transmission or reception).  This would help in supporting NTP-like capabilities to get civil time on Zephyr devices (something I'm working for Bluetooth mesh right now).

These use cases could be handled by using a second RTC peripheral at some (significant?) complexity required to align the clocks of the two RTCs, but it'd be nice to not waste those extra synchronized compare registers.

andyross commented:
> But what's preventing COUNTER from incrementing between the read of t2 and the set of the comparator on the next line?

Nothing.  But if that were fatal then it would never be possible to do this check robustly at all and the doucmentation would need to say 3 instead of 2 cycles.  And then software engineers would say **actually** we can't exactly detect 3 cycles, we need 4 to be safe.  And the docs would be updated, and....

You can't robustly lock against realtime.  All timing requirements like this have inherent race conditions.  If testing the counter and comparator on adjacent bus cycles like this isn't enough to meet the requirements of the specification, then nothing can be.  That doesn't mean the spec (or hardware) doesn't have bugs, but what this is doing is legal per paper.

In fact I played with this.  On nRF52, there is an internal pipeline/synchronizer/something delay of something like 1/5th of a RTC cycle.  As long as you get the value in earlier than that, the interrupt will set reliably.   You don't need 2 full cycles, ~1.2 is enough on this one platform.

As for the mechanism used to elevate to an ISR to trap a synchronous announcement, I'm neutral.  We don't need to use the NVIC, we can use a PendSV exception (though that adds a test to the userspace entry), set a flag to test in the interrupt exit code, call it synchronously underneath an existing ISR, trap to an SVC call in thread mode...  But this driver is written to use one comparator from one RTC, so this seemed like the natural mechanism.  It's certainly simple.

andyross commented:
A somewhat specious answer might help here: your concern could be addressed by simply setting the comparator first and then checking the counter to see if we "still had" two cycles left.  And maybe I should do that for clarity.

But as mentioned, the state returned by the RTC is dynamic, and the immediate counter read doesn't reflect the "still in flight" comparator write.  So really there's near-zero difference at the hardware level, and in fact the discrimination of the race condition that you really want isn't even possible in theory.  Which is why the docs pad the number in the first place :)

pabigot commented:
IMO you're making some unsafe assumptions how the peripheral actually behaves (including on nRF51, for which nRF52 observations may be a poor predictor).   But doing the read after the set would appear to address my concern, so that'd satisfy my change request.

andyross commented:
New version to address concerns.  Swapped the order of the comparator set and counter read for @pabigot.  Added a failsafe for features like ZLI that can preempt an irq lock for @pizi-nordic.

(Note that we have absolutely zero test coverage for CONFIG_ZERO_LATENCY_IRQS.  This was tested via some whiteboxed delays I stuffed in manually.  I'll file a bug on that.)

pfalcon commented on tests/posix/common/src/clock.c:
```diff
@@ -109,8 +110,16 @@ void test_posix_realtime(void)
+		 * (inclusive), but the sleep above is for 90ms.  A
+		 * high tick rate system will nail that delay exactly
+		 * and be right on the edge of a failure.  It should
+		 * be fixed (but not when the timer code is
```
> (but not when the timer code is
		 * simultaneously being worked on!).

I feel your pain, in PRs like https://github.com/zephyrproject-rtos/zephyr/pull/16621, https://github.com/zephyrproject-rtos/zephyr/pull/16427, https://github.com/zephyrproject-rtos/zephyr/pull/15937, our communication with guys like @nashif, @mbolivar is centered around me explaining (and them not understanding) that everything can't be fixed at once.

Then, let's still keep that to codereview, not code comments. If the nature of your patch is to replace `<`/`>` with `<=`/`>=` is timing comparison code, please do just that, and open a separate ticket with concerns expressed in code comments here (can be copied verbatim to that ticket).

pfalcon commented on tests/posix/common/src/clock.c:
```diff
@@ -109,8 +110,16 @@ void test_posix_realtime(void)
+		/* NOTE: this test is miscalibrated.  It's looking for
+		 * the clock to advance between 90 and 110ms
+		 * (inclusive), but the sleep above is for 90ms.  A
+		 * high tick rate system will nail that delay exactly
```
> A
		 * high tick rate system will

Your concern is a high tick rate system, and the concern of whoever wrote that test was a flaming qemu, where you sleep for 1ms, and get 10-20ms passed (though last time I checked, real hw wasn't much better than that). Your misunderstanding of motives behind that code as it is may come from that difference.

pfalcon reviewed (dismissed):
Some code comments should be turned into issue report instead.

andyross commented:
Removed the editorializing to address @pfalcon's concerns.

FWIW: I'm not a big fan of "file a bug to fix the test".  Like it or not, the tests define (more so even than the docs) the API we actually need to implement for real users.  In general my philosophy is that they should see minimal churn, even when we know they're of questionable quality.  The comment was there for future readers to understand why the test seems broken, and to guide reimplementations if we ever decide to throw it out.

pfalcon commented:
> Removed the editorializing to address @pfalcon's concerns.

Thanks, that commit looks much clearer now.

> Like it or not, the tests define (more so even than the docs) the API we actually need to implement for real users. In general my philosophy is that they should see minimal churn

I kinda agree, tests are kinda on meta-level to "normal" code. But in their own level, they're just software component with all its possible shortcomings and maintenance needs. And if we want to maintain it to best practices (fix issues consistently, and not in random-spot manner, etc.), documenting them as tickets and seek for wider exposure to gather critical mass to fix them well, is still the best way, like we or not.

Removing -1, but don't feel knowledgeable enough to give +1 for the entire patchset.

ruuddw reviewed (approved):
The ARC nsim simulator is currently configured to use instruction cycle accurate timers, not host OS timers; so it might actually work with the higher tick granularity as well - but, no real need for such accuracy on a simulator (and, didn't get to testing it now).

pizi-nordic commented:
Here are benchmark results on the updated PR.

k_timer() benchmark:

Difference between requested and measured timeout | master | #16398 | #16653 | #16782 
-|-|-|-|-
Minimum | -5189 us | 0 us | 0 us | 0 us
Average | 435.9 us | 23.6 us | 21.8 us | 22.1 us
Maximum | 7043 us | 59 us | 59 us | 60 us 

k_usleep() benchmark:

Difference between requested and measured timeout | master | #16398 | #16653 | #16782 
-|-|-|-|-
Minimum | 10179 us | 61 us | 30 us | 30 us 
Average | 11818.4 us | 112.0 us | 50.4 us | 56.8 us 
Maximum | 11856 us | 205 us | 88 us | 90 us 

Notes:
- Numbers presented for #16653 include overhead caused by counter abstraction as well as counter-generic layer.
- In order to compare apples with apples, the CONFIG_SYS_CLOCK_TICKS_PER_SEC has been set to 32768 for #16398.
- #16398 is not intended to improve timing of the OS API. Instead it proposes new k_alarm API, which is used to as clock source for kernel.

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -63,7 +61,10 @@ void rtc1_nrf_isr(void *arg)
+		/* As below: we're guaranteed to get an interrupt as
+		 * long as it's set two or more cycles in the future
+		 */
+		if (counter_sub(next, t) < 3) {
```
If CYC_PER_TICK = 1 (which is default) , the "next" may be still to close to current counter value for reliable interrupt.

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -117,21 +118,80 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+	 * work, the states overlap inside the counter hardware.  It's
+	 * possible to set a comparator value of "N", issue a DSB
+	 * instruction to flush the pipeline, and then immediately
+	 * read a counter value of "N-1" (i.e. the comparator is still
```
Side note: IMHO you can do the following:

```
set_comparator(cyc)
t = counter()
if  (counter_sub(cyc, t) >= 2) { OK interrupt will fire; } else { We have to do something. };
```

pizi-nordic commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -63,7 +61,10 @@ void rtc1_nrf_isr(void *arg)
+		/* As below: we're guaranteed to get an interrupt as
+		 * long as it's set two or more cycles in the future
+		 */
+		if (counter_sub(next, t) < 3) {
```
If CYC_PER_TICK = 1 (which is default) , the "next" may be still too close to current counter value for reliable interrupt.

IMHO we will need here something similar to the CONFIG_ZERO_LATENCY_IRQS workaround which you implemented below (next += 3).

andyross replied:
No, this is OK per docs.  As long as the time between the read of "t" above and this line are less than 30us (which is trivially verifiable by counting cycles).  The ZLI workaround is the failsafe for this too -- if it runs too long, we'll detect it there.

andyross replied:
And just to be clear: note we're masked here.  Nothing can interrupt us *except* a ZLI.  The timing under normal circumstances is 100% deterministic.

nordic-krch replied:
i think that same mechanism as in `z_clock_set_timeout` should be applied here, given ZLI presence.

pizi-nordic replied:
The ZLI is not only problem here. Since we are in the ISR and it not tickless, our `next` equals to `current + 1` for CYC_PER_TICK = 1. After correction, the next will be `current + 2` which is still too close to the current counter state value even if we assume that execution of the ISR taken no time.

IMHO we should make the code reliable (for example by scheduling next interrupt on `now +3`, not `now + 2^24+1`)  as well as put a warning here that we cannot have such high tick rate without tickless mode.

andyross replied:
> After correction, the next will be t + 2 which is still too close to the current counter value

It is not, though.  The docs for both nRF51[1] and nRF52[2] are quite clear that T+2 is guaranteed to produce an interrupt.

And like I mentioned earlier, I went in and (crudely) measured this.  It's not a synchronous device, the clock domains are different, and the overlap is in fact an analog thing and comes out to about 20% of a RTC cycle on nrf52840_pca10056.  The "+2" already includes the needed padding.

[1] "_nRF51 Series Reference Manual Version 3.0_", Section 19.1.7, p. 106

[2] "_nRF52840 Product Specification v1.0_" Section 6.22.7, p. 337

nordic-krch commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -20,8 +20,6 @@
 		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
 #define MAX_TICKS ((COUNTER_MAX - CYC_PER_TICK) / CYC_PER_TICK)
```
what if `CYC_PER_TICK = 1`? Then `MAX_TICKS = 0xffffff - 1`. Wouldn't it be simpler to set it to fixed `COUNTER_MAX/2`. Especially that there are some calculations there assuming that when subtraction gives result > 0x7fffff then counter is behind.

nordic-krch commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -117,21 +118,80 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+	} else {
+		set_comparator(cyc);
+		dt = counter_sub(cyc, counter());
+		if (dt == 0 || dt > 0x7fffff) {
```
imo, `MAX_TICKS` should be used here.

nordic-krch commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -20,8 +20,6 @@
-
 static struct k_spinlock lock;
 
 static u32_t last_count;
```
If it is possible that RTC CC is reconfigured before then there is a chance that previous alarm will generate interrupt. NOt sure if it is a problem here (it may mean only once additional interrupt without consequences). If so then we don't need to trouble else here is the algorithm i came up with to handle that (it is not that simple since event may be generated after changing CC (if it happens up to 15us before previous CC deadline).

https://github.com/zephyrproject-rtos/zephyr/blob/6bdbf47b2acf69533e26bb500b3a75f4ffb25085/drivers/counter/counter_nrfx_rtc.c#L188

nordic-krch commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -117,21 +118,80 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
+	 * that you need to be early by about 400 processor cycles
+	 * (about 1/5th of a RTC cycle) in order to reliably get the
+	 * interrupt.  The docs say two cycles, they mean two cycles.
+	 */
```
I would propose following algorithm. It would be common for ZLI and nonZLI case:
```
if (counter_sub(cyc, t) > 1) {
	set_comparator(cyc);
	t = counter();
   /* decrement to cover cyc == t which will return 0xffffff */
	dt = counter_sub(cyc - 1, t);
	if (dt > 0x7fffff) {
		/* EXPIRED, trigger irq */
	} else if (dt == 0) {
       /* cc is set to expire in one tick from now. it is possible that 
        * it was set too late thus move it 1 tick further */
		if (tick_delayed_alarm(c, t)) {
			/* EXPIRED, trigger irq */
		}
	} else {
		/* set on time, RTC interrupt will come */
	}
} else { /* now + 1 case, delay by one tick */
	if (tick_delayed_alarm(c, t)) {
		/* EXPIRED, trigger irq */
	}
}
```
and additional function for setting cc for value delayed by one tick:
```
static bool tick_delayed_alarm(u32_t c, u32_t now)
{
	set_comparator(c + 1);
	if (now != counter()) {
		/* 1 tick already elapsed */
		return true;
	}
	/* set on time, RTC interrupt will come */
	return false;
}
```

andyross replied:
This doesn't actually work.  That was the first trick I tried, actually.  The hardware states overlap.

It's possible to read a counter value of "N", set a comparator value of "N+1", immediately issue a DSB instruction to be sure the pipeline is flushed, and then read the counter back to see a "N".  You would think that the write got there in time, but it didn't and the interrupt won't arrive.

You have to be about 20% of a RTC cycle **ahead** of the transition to get the interrupt to trigger reliably, which the documentation pads to a full RTC cycle (i.e. you have to set N+2 to make it work).  That is what the current PR implements.

nordic-krch replied:
> It's possible to read a counter value of "N", set a comparator value of "N+1", immediately issue a DSB instruction to be sure the pipeline is flushed, and then read the counter back to see a "N". You would think that the write got there in time, but it didn't and the interrupt won't arrive.

True, but here we would be setting `now+2` since function is called when we see that cc `now+1`. In that case change to `n+2` and attempt to write. If after writing `now` has changed that means that 1 tick elapsed and we can say that requested timeout elapsed. Of course that is if 1 tick timeout means next tick and not `n+2` tick.

I've been testing the scenario where current CC has `now+1` value and we reconfigure to something in the future before CC value is reached. If we do that more than ~15us (half of 32k tick?) before the deadline then we won't get that interrupt, if later then old CC value will generate interrupt. So it seemed like CC is somehow latched between the ticks.

pizi-nordic reviewed (approved):
I am going on vacation, so I approve this PR in order to not block the merge during my absence.
@andyross: Please apply https://github.com/zephyrproject-rtos/zephyr/pull/16782#discussion_r296682519 or other mechanism which will ensure that we never set compare value too close to the current RTC value.

carlescufi commented:
@andyross I will leave it to you to decide whether to apply @nordic-krch 's suggestion in this PR or get this merge and then him follow-up on a second PR, since strictly speaking the suggestion is not directly related to the PR goals.

andyross commented:
I don't know that there's a huge rush.   This is an easy PR to rebase, so I'm happy to wait for debate to finish.

carlescufi commented:
> I don't know that there's a huge rush. This is an easy PR to rebase, so I'm happy to wait for debate to finish.

Not *huge* but we definitely need this for 2.0 (cc @ioannisg)

nordic-krch commented:
@carlescufi 
> @andyross I will leave it to you to decide whether to apply @nordic-krch 's suggestion in this PR or get this merged and then him follow-up on a second PR

I think that can be fixed later on.

carlescufi commented:
@andyross could you rebase so we can merge?

carlescufi commented:
Still some doxygen errors @andyross, not sure why those suddenly appeared now.

dbkinder commented on include/kernel.h:
```diff
@@ -1708,9 +1707,8 @@ void k_disable_sys_clock_always_on(void);
- *    default setting of 100, system time is updated in increments of 10ms.
- *    @endrst
+ *    :option:`CONFIG_SYS_CLOCK_TICKS_PER_SEC` config option
+ *    @endrststar
```
doxygen brackets for embedded restructured text are now simply ``@rst`` ... ``@endrst``  (we got rid of the @endrststar alias to simplify things).  it was correct as ``@endrst`` in the previous version of this file.

dbkinder reviewed (changes requested):
use @rst and @endrst to bracket embedded restructuredText in doxygen comments (that will fix the doc build error here)

andyross commented:
Rebased.   Fixed up doc errors due to the endrst/endrststar change that this crossed and mismerged with.
